### PR TITLE
Operator `__pragma`

### LINKED ISSUE #267: __attribute__((attr-list)) or [[attr-list]]
`#pragma` is a preprocessor directive; hence, apart from being ugly, it puts limitations on its usage. For example, pragma cannot be used inside substitution pattern of a `#define` (`of a  ??`; what is the correct technical term?).

If something like the `__attribute__(())` in GCC or like `[[]]` in C++17 (or `__declspec`  in Microsoft; not symmetric and not as readable as the other two in my opinion) is added, this would get rid of those limitations and also improve readability of the code.

# Design ideas

## Basic Examples:
Some the attributes used in this section do not exist in the current compiler. I have used them to just present the possible syntax so that viewers can get an idea how it would look when used in code (some cases such as statement attributes do not have any options that could be used in the current version of the compiler but they could be added in the future so let's consider those as well).

 - function attributes
`stock __attribute__((naked)) myfunc () { }` 
`stock myfunc () __attribute__((naked)) { }` 
`[[naked]] stock myfunc () { }` 
`stock myfunc () [[naked]] { }` 

 - function argument attributes
`stock myfunc(__attribute__((unused)) myarg, normal_arg) { } `
`stock myfunc([[unused]] myarg, normal_arg) { }`
`stock myfunc(myarg __attribute__((unused)) , normal_arg) { } `
`stock myfunc(myarg [[unused]] , normal_arg) { }`

 - variable attributes
`new __attribute__((unused)) var;`
`new [[unused]] var;`
`new var __attribute__((unused));`
`new var [[unused]] ;`

 - label attributes
`__attribute__((omit-defaults)) mylabel:` (potential solution for a part of #266 )
`[[omit-defaults]] mylabel:` 
`mylabel: __attribute__((omit-default));`
`mylabel: [[omit-defaults]];` 

 - enumerator attributes
 ```
 enum {
       __attribute__((deprecated)) score,
       [[deprecated]] money,
       points __attribute__((deprecated)),
       something [[depreacted]]
 };

enum __attribute__((unused)) { }
enum [[unused]] { }
 ```
 
- statement attributes
`__attribute__((nodestruct)) return 0;`
`[[nodestruct]] return 0;`

## Empty attributes:
`stock __attribute__(()) myfunc() { }`
`new __attribute__(()) myvar;`

This might provide more flexibility when attributes are a part of substitution pattern.

## Multiple attributes:

`stock __attribute__((naked)) __attribute__((deprecated))  myfunc () {}` 
`[[naked]] [[deprecated]] stock  myfunc () {}` 

## Multiple options:

`stock __attribute__((naked, deprecated)) myfunc() {}`
`[[naked, deprecated]] stock  myfunc() {}`

## Attribute parameters:
`stock __attribute__((naked(32))) myfunc () {}` 
`[[naked(32)]] stock myfunc () {}` 

`stock __attribute__((deprecated("because bla"))) myfunc () {}` 
`[[deprecated("because bla")]] stock myfunc () {}`

### PR DESCRIPTION
**Pull request by Daniel-Cortez**
**What this PR does / why we need it**:

This PR implements operator `__pragma` (see #267).

Currently available options:
* `deprecated`
* `unused`
* `unread`
* `unwritten`
* `nodestruct`
* `naked`

Examples of use:
```Pawn
native Func() __pragma("deprecated - use OtherFunc() instead");

stock __pragma("naked") NakedFunc2() {}

NakedFunc()
{
    __pragma("naked", "deprecated - use NakedFunc2() instead");
}

new __pragma("nodestruct") Tag:b = Tag:0;
new a = 0 __pragma("unused");
new __pragma("unread") c = 0 __pragma("unwritten");
new d = 0 __pragma("unread", "unwritten");
```

**Which issue(s) this PR fixes**:

Fixes #267

**What kind of pull this is**:

* [ ] A Bug Fix
* [x] A New Feature
* [ ] Some repository meta (documentation, etc)
* [ ] Other

**Additional Documentation**:

### CODE REVIEW

**Comment by Y-Less** [comment]
> `__pragma("unread", "unwritten");`

That's a good solution to multiple properties.

**Comment by Y-Less** [comment]
Why didn't you reuse the existing `#pragma` code?  It seems it would have been simpler to just use that and add an extra parameter to disable some like `codepage` for `__` compared to `#`.

**Comment by Daniel-Cortez** [comment]
Because the code for `__` is too different from `#`.

* The option names are handled differently.
Since operator `__pragma` can be specified before the symbol is known, e.g.
```
new __pragma("nodestruct") Tag:var;
```
  ... , I had to make the compiler memoize the specified attributes, so they could be applied to the symbol later - the most obvious solution was to store bit flags into a global variable.
This is why while `#pragma` distinguishes the options by their names, `__pragma` uses bit flags when applying the attributes.
Of course, I could just convert the bit flags back into attribute names, but there are still other problems listed below.

* The option names and parameters are read differently.
In `#pragma` the option name and parameters are read directly (by manipulating with `lptr`), but in `__pragma` they are read from the literal queue, because the operator reuses the standard string parsing mechanism.

* The parameters are handled differently.
In `#pragma unused/unread/unwritten/nodestruct` the symbol name is specified after the option name (e.g. `#pragma nodestruct var`), but in `__pragma` the symbol is separate from the operator.

Thus the only thing left to reuse is the code responsible for applying the attributes to symbols (which isn't really that much of code), which could be moved into separate reusable functions:
```C
SC_FUNC void pragma_unused(symbol *sym, int unread, int unwritten)
{
  sym->usage |= unwritten ? 0 : uREAD;
  if (sym->ident == iVARIABLE || sym->ident == iREFERENCE
      || sym->ident == iARRAY || sym->ident == iREFARRAY)
    sym->usage |= unread ? 0 : uWRITTEN;
}

SC_FUNC void pragma_nodestruct(symbol *sym)
{
  if (sym->ident==iVARIABLE || sym->ident==iARRAY)
    sym->usage |= uNODESTRUCT;
}
```

**Comment by Y-Less** [comment]
That makes sense.

**Comment by Daniel-Cortez** [comment]
Done, moved the code for `deprecated`, `unused`, `unread`. `unwritten` and `nodestruct` into reusable functions, as described in the post above.

While I was at it, I also fixed a potential memory leak in `#pragma deprecated` (the memory block for the deprecaton message is not freed if it's unused, which might happen either because of end of script or if there are two deprecation `#pragma`s used in a row).

**Comment by Y-Less** [comment]
I've merged this, but it would be nice if it were possible to use `"naked"` inside functions as well as in their header:

```pawn
Func()
{
    __pragma("naked");
}
```

**Comment by Daniel-Cortez** [comment]
> it would be nice if it were possible to use `"naked"` inside functions as well as in their header

Done.
As a bonus, aside from `"naked"`, the other attributes can be used inside functions as well:
```Pawn
NakedFunc()
{
	__pragma("naked", "deprecated - use NakedFunc2() instead");
}
```

**Comment by Y-Less** [comment]
Oh, I just discovered that `naked` removes `PROC` as well.  That's good to know.

**Comment by Y-Less** [comment]
The tests don't really cover that, but they're only testing compiler warnings.  Is that actually correct?

**Comment by Daniel-Cortez** [comment]
> The tests don't really cover that, but they're only testing compiler warnings. Is that actually correct?

If by "that" you mean all other attributes not being tested inside functions, then I think it might unnecessarily bloat the tests. As you can see it in ab66969269f6f7e3834706f8608a8e42235fbdea, the only difference is in the code that calls functions `dopragma()` and `pragma_apply(curfunc)`, making it possible for `__pragma` to be used inside functions; the code that reads and applies the attributes is still the same, which is why I added only one basic check to make sure the syntax is being handled correctly. But I can expand the tests if you think that check isn't enough.

**Comment by Y-Less** [comment]
Let me rephrase:

I just discovered that `naked` removes `PROC` as well, is that correct?  The tests only cover compiler warnings, not assembly output, so they don't cover the lack of `PROC`.

I was just talking about `PROC` in both posts.

**Comment by Daniel-Cortez** [comment]
Ah, so that's what you meant.
Alright then, I'll add a P-code test.

EDIT: OK, done: #571.
